File: test/test_duo_conformance.jl
@testsnippet DuoConformanceSetup begin
    using Test
    using EarthSciDiscretizations
    using JSON

    # Julia's DuoGrid is 1-indexed and uses `0` for "no neighbor"; the
    # binding-neutral golden stores 0-based cell indices and `-1` for
    # boundary. Convert at the harness boundary so all four bindings
    # compare byte-for-byte on the same numbers.
    _neighbors_0idx(g, c1::Int) = [
        n == 0 ? -1 : n - 1 for n in EarthSciDiscretizations.neighbors(g, c1)
    ]
end

@testitem "DUO cross-language conformance" setup = [DuoConformanceSetup] tags = [:conformance, :grid, :duo] begin
    HARNESS_DIR = joinpath(@__DIR__, "..", "tests", "conformance", "grids", "duo")
    FIXTURES = JSON.parsefile(joinpath(HARNESS_DIR, "fixtures.json"))
    REL_TOL = Float64(FIXTURES["tolerance"]["relative"])
    METRIC_NAMES = (:area, :lon, :lat, :x, :y, :z)

    function close_rel(a::Real, b::Real, tol::Real)
        scale = max(1.0, abs(a), abs(b))
        return abs(a - b) <= tol * scale
    end

    for fixture in FIXTURES["fixtures"]
        name = fixture["name"]
        opts = fixture["opts"]
        level = Int(opts["level"])
        R = float(opts["R"])
        ghosts = Int(opts["ghosts"])
        golden = JSON.parsefile(joinpath(HARNESS_DIR, "golden", "$name.json"))

        @testset "fixture=$name" begin
            loader = DuoLoader(path = "builtin://icosahedral/$level")
            grid = build_duo_grid(; loader = loader, R = R, ghosts = ghosts)

            @test n_cells(grid) == golden["n_cells"]
            @test n_vertices(grid) == golden["n_vertices"]
            @test n_edges(grid) == golden["n_edges"]

            qps = Int.(fixture["query_points"])
            @test length(qps) == length(golden["cell_centers"])

            for (k, c0) in enumerate(qps)
                c1 = c0 + 1

                lon, lat = cell_centers(grid, c1)
                gc = golden["cell_centers"][k]
                @test close_rel(lon, gc["lon"], REL_TOL)
                @test close_rel(lat, gc["lat"], REL_TOL)

                got_nb = _neighbors_0idx(grid, c1)
                expected_nb = [Int(x) for x in golden["neighbors"][k]]
                @test got_nb == expected_nb

                for m in METRIC_NAMES
                    got = metric_eval(grid, m, c1)
                    expected = Float64(golden["metric_$(m)"][k])
                    @test close_rel(got, expected, REL_TOL)
                end
            end
        end
    end
end

File: python/tests/test_duo_conformance.py
"""Cross-language conformance for the DUO grid family.

Verifies that the Python accessor runtime produces accessor outputs that
match the committed Julia-reference golden within the documented
``docs/GRIDS_API.md`` §4.2 ULP tolerance. Corpus lives at
``tests/conformance/grids/duo/``.
"""

from __future__ import annotations

import json
from pathlib import Path

import pytest

from earthsci_discretizations import grids

HARNESS_DIR = (
    Path(__file__).resolve().parents[2]
    / "tests"
    / "conformance"
    / "grids"
    / "duo"
)
FIXTURES_PATH = HARNESS_DIR / "fixtures.json"
GOLDEN_DIR = HARNESS_DIR / "golden"
_SPEC = json.loads(FIXTURES_PATH.read_text())
REL_TOL = float(_SPEC["tolerance"]["relative"])
_FIXTURE_NAMES = [f["name"] for f in _SPEC["fixtures"]]

_METRIC_NAMES = ("area", "lon", "lat", "x", "y", "z")


def _close_rel(a: float, b: float, tol: float = REL_TOL) -> bool:
    scale = max(1.0, abs(a), abs(b))
    return abs(a - b) <= tol * scale


def _fixture(name: str) -> dict:
    for f in _SPEC["fixtures"]:
        if f["name"] == name:
            return f
    raise KeyError(name)


@pytest.mark.parametrize("fixture_name", _FIXTURE_NAMES)
def test_duo_conformance_against_golden(fixture_name: str) -> None:
    fixture = _fixture(fixture_name)
    opts = fixture["opts"]
    level = int(opts["level"])
    grid = grids.duo(
        loader={"path": f"builtin://icosahedral/{level}"},
        R=float(opts["R"]),
        ghosts=int(opts["ghosts"]),
        dtype=opts["dtype"],
    )
    golden = json.loads((GOLDEN_DIR / f"{fixture_name}.json").read_text())

    assert grid.n_cells == golden["n_cells"]
    assert grid.n_vertices == golden["n_vertices"]
    assert grid.n_edges == golden["n_edges"]
    assert len(fixture["query_points"]) == len(golden["cell_centers"])

    for k, c in enumerate(fixture["query_points"]):
        c = int(c)
        lon, lat = grid.cell_centers(c)
        gc = golden["cell_centers"][k]
        assert _close_rel(lon, gc["lon"]), f"lon mismatch at qp[{k}]=c{c}"
        assert _close_rel(lat, gc["lat"]), f"lat mismatch at qp[{k}]=c{c}"

        got_nb = list(grid.neighbors(c))
        expected_nb = [int(x) for x in golden["neighbors"][k]]
        assert got_nb == expected_nb, f"neighbors mismatch at qp[{k}]=c{c}"

        for m in _METRIC_NAMES:
            got = grid.metric_eval(m, c)
            expected = float(golden[f"metric_{m}"][k])
            assert _close_rel(got, expected), (
                f"metric {m} mismatch at qp[{k}]=c{c}: got {got}, expected {expected}"
            )
